lib/OpenQA/WebAPI/Plugin/ObsRsync/Gru.pm

```diff
@@ -0,0 +1,179 @@
+    }
+
+    $self->render('ObsRsync_gru_index', jobs => \%jobs);
+}
+
+# this prepares fields for rendering
+sub _extend_job_info {
+    my ($job) = @_;
+
+    my $created_at = $job->{created};
+    $created_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($created_at)) if $created_at;
+
+    my $started_at = $job->{started};
+    $started_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($started_at)) if $started_at;
+
+    my $args = $job->{args};
+    if (ref $args eq 'HASH' && scalar(%$args) == 1 && $args->{project}) {
+        $args = $args->{project};
+    }
+    else {
+        $args = dump($args);
+    }
+    my $info = {
+        id       => $job->{id},
+        task     => $job->{task},
+        args     => $args,
+        created  => $created_at,
+        started  => $started_at,
+        priority => $job->{priority},
+        retries  => $job->{retries},
+        state    => $job->{state},
+        notes    => dump($job->{notes}),
+    };
+    return $info;
+}
+
+sub run {
+    my $self    = shift;
+    my $project = $self->param('folder');
+    return undef if $self->_check_and_render_error($project);
+    my $app         = $self->app;
+    my $queue_limit = $app->obs_rsync->queue_limit;
+
+    my $results = $app->minion->backend->list_jobs(
+        0,
+        4 * $queue_limit,
+        {tasks => ['obs_rsync_run', 'obs_rsync_queue'], states => ['active', 'inactive']});
+
+    for my $other_job (@{$results->{jobs}}) {
+        if ($other_job->{args} && ($other_job->{args}[0]->{project} eq $project)) {
+
+            return $self->render(json => {message => $project . ' already in queue'}, status => IN_QUEUE)
+              if ($other_job->{task} eq 'obs_rsync_queue' || $other_job->{notes}{waitingconcurrencyslot});
+            $app->gru->enqueue('obs_rsync_queue', {project => $project}, {priority => 90});
+            return $self->render(json => {message => "ok"}, status => QUEUED);
+        }
+    }
+    return $self->render(json => {message => 'queue full'}, status => QUEUE_FULL)
+      if ($results->{total} >= $queue_limit);
+
+    $app->gru->enqueue(
+        'obs_rsync_run',
+        {project  => $project},
+        {priority => 100, notes => {waitingconcurrencyslot => 1}});
+    return $self->render(json => {message => 'started'}, status => STARTED);
+}
+
+sub _queue {
+    my ($app, $job, $args) = @_;
+    my $project        = $args->{project};
+    my $retry_interval = $app->obs_rsync->retry_interval;
+    my $queue_limit    = $app->obs_rsync->queue_limit;
+
+
+    my $results = $app->minion->backend->list_jobs(
+        0,
+        4 * $queue_limit,
+        {tasks => ['obs_rsync_run'], states => ['active', 'inactive']});
+    for my $other_job (@{$results->{jobs}}) {
+        if ($other_job->{args} && ($other_job->{args}[0]->{project} eq $project)) {
+            return $job->finish(IN_QUEUE) if $other_job->{notes}{waitingconcurrencyslot};
+            return $job->retry({delay => $retry_interval});
+        }
+    }
```

Comment:
This part still worries me a little. I don't quite understand what it is meant to do. Where is this covered in the tests?

Reply:
It is executed in Gru worker, so it is tricky to show in cover properly. But it must be executed in async test: 'run' job is being executed 2 sec, and retry is configured 1 sec, so it triggers at least once for each job in queue https://github.com/os-autoinst/openQA/pull/2225/files#diff-f63eac5cc88e85f458d7840e440e5877R34